provider/query_parameter_value.go

```diff
@@ -0,0 +1,52 @@
+package provider
+
+import (
+	"fmt"
+	"strings"
+)
+
+// Query parameter holds normalized parameter data ready to be inserted in the
+//  final query
+type QueryParameterValue struct {
+	// Token to replace e.g., !TOKEN!
+	Token string
+	// SQL expression to be inserted. Contains "?" that will be replaced with an
+	//  ordinal argument e.g., "$1"
+	SQL string
+	// Value that will be passed to the final query in arguments list
+	Value interface{}
+	// Raw parameter values for debugging and monitoring
+	RawValues map[string]string
+}
+
+// ReplaceParams substitutes configured query parameter tokens for their values
+//  within the provided SQL string
+func ReplaceParams(params map[string]QueryParameterValue, sql string, args *[]interface{}) string {
+	if params == nil {
+		return sql
+	}
+
+	for _, token := range ParameterTokenRegexp.FindAllString(sql, -1) {
```

Comment:
I assume that at the point when `ReplaceParams` is called, all built-in tokens have already been replaced?

Reply:
Yes, but shouldn't matter. After finding a token in the incoming SQL we look for it in the list of parameter tokens and then replace. I will add a check to ensure that the token was indeed found.